contracts/bonding/BondingManager.sol

```diff
@@ -307,6 +336,11 @@ contract BondingManager is ManagerProxyTarget, IBondingManager {
                 totalStake,
                 currentRoundTotalActiveStake
             );
+
+            // Deduct what would have been the treasury rewards
+            uint256 treasuryRewards = MathUtils.percOf(rewards, treasuryRewardCutRate);
+            rewards = rewards.sub(treasuryRewards);
```

Comment:
Might need some special attention here, as I had trouble understanding what is really the purpose of the whole logic in this `if` branch.

I understand _when_ it is called, but it seems like it is only capable of making an approximation of what the previous round rewards would have been, but it's not an exact calculation. 

On line 335 it also uses `mintableTokens+mintedTokens` as the estimate for the "total previous round rewards", when from my understanding it should be only `mintableTokens` assuming the previous round had similar mintable tokens as the current round. Is there any similar LIP or post explaining the mathematics done here?

So there's also a theoretical problem here as well, which is that we always use the current cut rate to calculate what the previous round treasury cut would have been. This can be theoretically a problem if the cut rate has changed since the last round, but it seems like this is too much of a corner case + already an approximation for us to need any code complexity to make this more exact.

WDYT?

Reply:
A couple resources:

- [LIP-36 section on fees - search for the phrase "retroactively calculate"](https://github.com/livepeer/LIPs/blob/master/LIPs/LIP-36.md#fees)
- [Original comment discussing the retroactive calculation](https://github.com/livepeer/LIPs/issues/35#issuecomment-673659199)
- [Cumulative Reward and Fee Factors for Delegators section of LIP-36 discussion thread](https://forum.livepeer.org/t/a-more-gas-efficient-earnings-calculation-approach/1097)

IIRC the logic here is:

- `cumulativeRewardFactor(currentRound) = cumulativeRewardFactor(currentRound - 1) * (1 + delegatorsRewards(currentRound) / transcoderActiveStake(currentRound)`
- Suppose we know `cumulativeRewardFactor(currentRound)`, but we do not know `cumulativeRewardFactor(currentRound - 1)`
- We can rearrange the first equation to get: `cumulativeRewardFactor(currentRound - 1) = cumulativeRewardFactor(currentRound) / (1 + delegatorsRewards(currentRound) / transcoderActiveStake(currentRound)`

> On line 335 it also uses mintableTokens+mintedTokens as the estimate for the "total previous round rewards"

This is supposed to be the total mintable tokens for the current round which is used to derive the rewards that the transcoder minted in the current reward in a reward call via the equation: `rewards = total mintable tokens * (transcoderActiveStake(currentRound) / totalActiveStake(currentRound))`. Then, `delegatorsRewards(currentRound)` can be derived from `rewards` by subtracting the transcoder's reward cut.

> So there's also a theoretical problem here as well, which is that we always use the current cut rate to calculate what the previous round treasury cut would have been.

Since we're trying to derive `delegatorsRewards(currentRound)` as long as `treasuryRewardCutRate` does not change in between the transcoder's reward call in the current round and the time at which this logic is executed we avoid a problem...

A few options to avoid that scenario:

1. Protocol governance should only execute an update that changes `treasuryRewardCutRate` towards the end of a round so that by the time reward calls in the next round occur the `treasuryRewardCutRate` is locked in and won't change after that
2. Store additional state to track `treasuryRewardCutRate` for each round so that even if there is an update can always use the stored value for the round
3. Store an extra `nextTreasuryRewardCutRate`. `setTreasuryRewardCutRate()` updates `nextTreasuryRewardCutRate` and then `setCurrentRoundTotalActiveStake()` also sets `treasuryRewardCutRate = nextTreasuryRewardCutRate` which means that when a round is initialized (and `setCurrentRoundTotalActiveStake()` is called) the `treasuryRewardCutRate` is locked in for the round

Slight pref for 3 as it seems to be the cleanest approach. WDYT?

Reply:
Thanks for the references! I believe I understand the overall need for this logic. In practice, it does not make an estimate but calculates exactly the value of `cumulativeRewardFactor` for the previous reward round. This _should_ result in exactly the value on the earning pool on the previous reward call. 

In a sense, this could've been solved with the checkpointing system that we are building for this. IOW if we had a field like `previousRewardRound` inside every earnings pool, we could fetch the value directly from that last round where reward() was run before the current one and this logic wouldn't be needed.

There is still something that looks weird to me though. Specifically:
 - [This line](https://github.com/livepeer/protocol/blob/cf12cb4734dc1e7e734faed18513f38502d61095/contracts/bonding/BondingManager.sol#L340) calculates the `rewards` value, which are the reward tokens distributed to this transcoder in the current round. It uses `currentMintableTokens + currentMintedTokens` to calculate the total round rewards and take the perc of
 - OTOH on the actual reward call, we calculate this via the [`createReward()`](https://github.com/livepeer/protocol/blob/cf12cb4734dc1e7e734faed18513f38502d61095/contracts/bonding/BondingManager.sol#L889) function on the minter which actually only uses [`currentMintableTokens`](https://github.com/livepeer/protocol/blob/d1e2222173618cb5cd905fd58bb78c72e521b20f/contracts/token/Minter.sol#L153) for the "total round rewards" and take the same perc of
 
So it looks like the first one might be calculating them wrong, inflating the value of what the transcoder rewards were in the current round. This doesn't affect the treasury reward logic tho so doesn't block this PR, but if there is indeed a problem there I could include the fix on this changeset (basically remove the `+ currentMintedTokens`).

Will send a separate msg about the `treasuryReardCutRate` itself

Reply:
Will follow up offline!